scipy/optimize/_minimize.py

```diff
@@ -601,54 +601,99 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
                       bounds=bounds, constraints=constraints,
                       callback=callback, **options)
 
-    if bounds is not None:
-        bounds = standardize_bounds(bounds, x0, meth)
-
     if constraints is not None:
         constraints = standardize_constraints(constraints, x0, meth)
 
+    remove_vars = False
+    if bounds is not None:
+        if meth.upper() in {"TNC", "SLSQP", "L-BFGS-B"}:
+            # These methods can't take the finite-difference derivatives they
+            # need when a variable is fixed by the bounds. To avoid this issue,
+            # remove fixed variables from the problem.
+
+            # convert to new-style bounds so we only have to consider one case
+            bounds = standardize_bounds(bounds, x0, 'new')
+
+            # determine whether any variables are fixed
+            i_fixed = (bounds.lb == bounds.ub)
+
+            # determine whether finite differences are needed for any grad/jac
+            fd_needed = not(callable(jac))
+            for con in constraints:
+                if not callable(con.get('jac', None)):
+                    fd_needed = True
+
+            # if F.D. are ever used, remove all fixed variables
+            remove_vars = i_fixed.any() and fd_needed
+            if remove_vars:
+                x_fixed = (bounds.lb)[i_fixed]
+                x0 = x0[~i_fixed]
+                bounds = _remove_from_bounds(bounds, i_fixed)
+                fun = _remove_from_func(fun, i_fixed, x_fixed)
+                if callable(callback):
+                    callback = _remove_from_func(callback, i_fixed, x_fixed)
+                if callable(jac):
```

Comment:
This is to consider the case where the gradient (called `jac`, unfortunately) is callable but constraint jacobian (`con['jac']`) requires finite differences and vice versa : / Unfortunate that we have to do this.